src/util/range/numeric_range.c

```diff
@@ -0,0 +1,125 @@
+/*
+* Copyright 2018-2019 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "numeric_range.h"
+#include "../rmalloc.h"
+#include "../../ast/ast_shared.h"
+#include <math.h>
+#include <assert.h>
+
+NumericRange *NumericRange_New(void) {
+	NumericRange *range = rm_malloc(sizeof(NumericRange));
+	range->valid = true;
+	range->max = INFINITY;
+	range->min = -INFINITY;
+	range->include_min = false;
+	range->include_max = false;
+	return range;
+}
+
+bool NumericRange_IsValid(const NumericRange *range) {
+	if(range->include_min && range->include_max) {
+		// X >= y AND X <= z
+		return (range->min <= range->max);
+	} else {
+		// X >= y AND X < z
+		// X > y AND X <= z
+		// X > y AND X < z
+		return (range->min < range->max);
+	}
+}
+
+bool NumericRange_ContainsValue(const NumericRange *range, double v) {
+	if(!range->valid) return false;
+
+	// Make sure v is <= max.
+	if(range->include_max) {
+		if(v > range->max) return false;
+	} else {
+		if(v >= range->max) return false;
+	}
+
+	// Make sure v >= min.
+	if(range->include_min) {
+		if(v < range->min) return false;
+	} else {
+		if(v <= range->min) return false;
+	}
+
+	return true;
+}
+
+void NumericRange_TightenRange(NumericRange *range, int op, double v) {
+	if(!range->valid) return;
+
+	switch(op) {
+	case OP_LT:    // <
+		if(range->max >= v) {
+			range->include_max = false;
+			range->max = v;
+		}
+		break;
+	case OP_LE:    // <=
+		if(range->max > v) {
+			range->include_max = true;
+			range->max = v;
+		}
+		break;
+	case OP_GT:    // >
+		if(range->min <= v) {
+			range->include_min = false;
+			range->min = v;
+		}
+		break;
+	case OP_GE:    // >=
+		if(range->min < v) {
+			range->include_min = true;
+			range->min = v;
+		}
+		break;
+	case OP_EQUAL:  // =
```

Comment:
Misbehaves on multiple equality operations:
```
$ ge social "MATCH (a:person) WHERE a.age = 100 AND a.age = 30 RETURN a"
1) "Results"
2) "    Project"
3) "        Index Scan | (a:person)"

gq social "MATCH (a:person) WHERE a.age = 100 AND a.age = 30 RETURN a"
1) 1) "a"
2) 1) 1) 1) 1) "id"
            2) (integer) 23
         2) 1) "labels"
            2) 1) "person"
         3) 1) "properties"
            2) 1) 1) "status"
                  2) "married"
               2) 1) "gender"
                  2) "female"
               3) 1) "age"
                  2) (integer) 30
               4) 1) "name"
                  2) "Lucy Yanfital"
```

Reply:
👍 bug in range validation, now in case the range is invalid we use RediSearch empty QueryNode.